In [1]:
import mysql.connector
from mysql.connector import Error
import pandas as pd
from datetime import date

In [2]:
def create_server_connection(host_name, user_name, user_password):
    connection = None
    try:
        connection = mysql.connector.connect(
            host=host_name,
            user=user_name,
            passwd=user_password
        )
        print("MySQL Database connection successful")
    except Error as err:
        print(f"Error: '{err}'")

    return connection

pw = "dbgush00" # IMPORTANT! Put your MySQL Terminal password here.

connection = create_server_connection("localhost", "root", pw)

MySQL Database connection successful


In [3]:
def create_database(connection, query):
    cursor = connection.cursor()
    try:
        cursor.execute(query)
        print("Database created successfully")
    except Error as err:
        print(f"Error: '{err}'")

#create_database_query = "CREATE DATABASE plataformacapital"
#create_database(connection, create_database_query)

## Funções uteis para banco de dados

In [4]:
def create_db_connection(host_name, user_name, user_password, db_name):
    connection = None
    try:
        connection = mysql.connector.connect(
            host=host_name,
            user=user_name,
            passwd=user_password,
            database=db_name
        )
        print("MySQL Database connection successful")
    except Error as err:
        print(f"Error: '{err}'")

    return connection

In [5]:
def execute_query(connection, query):
    cursor = connection.cursor()
    try:
        cursor.execute(query)
        connection.commit()
        print("Query successful")
    except Error as err:
        print(f"Error: '{err}'")

In [6]:
def read_query(connection, query):
    cursor = connection.cursor()
    result = None
    try:
        cursor.execute(query)
        result = cursor.fetchall()
        return result
    except Error as err:
        print(f"Error: '{err}'")


### Criando as tabelas

In [7]:
db = "plataformacapital" # This is the name of the database we will create in the next step - call it whatever you like.

In [8]:
create_t25_table = """
CREATE TABLE tbng1025_vers_agem_flux_capi (
  cod_agem_flux_capi INT,
  num_vers_agem_flux_capi INT,
  cod_situ_exeo_flux_capi INT,
  cod_tipo_esra_capi INT,
  ind_utiz_vers_agem VARCHAR(1),
  dat_rfrc_psst DATE
  );
 """

create_t21_table = """
CREATE TABLE tbng1021_situ_exeo_flux_capi (
  cod_situ_exeo_flux_capi INT PRIMARY KEY,
  nom_tipo_base_orig_capi VARCHAR(256)
);
 """

create_t4_table = """
CREATE TABLE tbng1004_agem_flux_capi (
  cod_agem_flux_capi INT PRIMARY KEY,
  nom_agem_flux_capi VARCHAR(256)
);
"""

create_t24_table = """
CREATE TABLE tbng1024_tipo_esra_capi (
  cod_tipo_esra_capi INT PRIMARY KEY,
  nom_tipo_esra_capi VARCHAR(256)
);
"""

create_t3_table = """
CREATE TABLE tbng1003_agem_cenr_flux_capi (
  cod_agem_flux_capi INT,
  num_vers_agem_flux_capi INT,
  cod_cenr_flux_capi INT,
  num_vers_cenr_flux_capi INT
);
"""

create_t13_table = """
CREATE TABLE tbng1013_cenr_prso_flux_capi (
  cod_cenr_flux_capi INT,
  num_vers_cenr_flux_capi INT,
  cod_prso_flux_capi INT,
  num_vers_prso_flux_capi INT,
  cod_situ_exeo_flux_capi INT,
  dat_inio_psst DATE,
  dat_fim_psst DATE
);
"""

create_t99_table = """
CREATE TABLE tbng1099_arqu_vrto_cenr_prso (
  cod_cenr_flux_capi INT,
  num_vers_cenr_flux_capi INT,
  cod_prso_flux_capi INT,
  num_vers_prso_flux_capi INT,
  nom_arqu_vrto VARCHAR(50),
  dat_rfrc_vrto DATE,
  num_vrto_arqu INT
);
"""

create_t17_table = """
CREATE TABLE tbng1017_prso_base_flux_capi (
  cod_prso_flux_capi INT,
  num_vers_prso_flux_capi INT,
  cod_base_orig_capi INT
);
"""

create_t2_table = """
CREATE TABLE tbng1002_vrto_recp_arqu_capi (
  nom_arqu_vrto_recp_arqu_capi VARCHAR(50),
  num_vrto_arqu INT,
  dat_rfrc_vrto_arqu DATE,
  dat_recm_vers_arqu DATE,
  cod_base_orig_capi INT
);
"""


connection = create_db_connection("localhost", "root", pw, db)
execute_query(connection, create_t25_table) # Execute our defined query
execute_query(connection, create_t21_table)
execute_query(connection, create_t4_table)
execute_query(connection, create_t24_table)
execute_query(connection, create_t3_table)
execute_query(connection, create_t13_table)
execute_query(connection, create_t99_table)
execute_query(connection, create_t17_table)
execute_query(connection, create_t2_table)

MySQL Database connection successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful


### Populando as tabelas

In [42]:
lista_tabelas = ['tbng1025_vers_agem_flux_capi','tbng1021_situ_exeo_flux_capi','tbng1004_agem_flux_capi',
                'tbng1024_tipo_esra_capi','tbng1003_agem_cenr_flux_capi','tbng1013_cenr_prso_flux_capi',
                'tbng1099_arqu_vrto_cenr_prso','tbng1017_prso_base_flux_capi','tbng1002_vrto_recp_arqu_capi']
for tabela in lista_tabelas:
    execute_query(connection, f"TRUNCATE TABLE {tabela}")

Query successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful


In [43]:
pop_t25 = """
INSERT INTO tbng1025_vers_agem_flux_capi VALUES
(0, 100, 0, 7001, 'P', '1968-07-20'),
(1, 101, 2, 7001, 'P', '1985-08-20'),
(2, 102, 1, 7002, 'P', '1985-05-20'),
(3, 103, 0, 7001, 'S', '2000-04-22'),
(4, 104, 2, 7002, 'P', '2011-04-20'),
(5, 105, 2, 7001, 'P', '2020-04-20'),
(6, 106, 0, 7001, 'S', '2022-04-20'),
(7, 107, 1, 7003, 'S', '2022-04-20'),
(8, 108, 0, 7001, 'S', '2021-04-23'),
(9, 109, 0, 7003, 'P', '2021-04-22');
"""

pop_t24 = """
INSERT INTO tbng1024_tipo_esra_capi VALUES
(7001, 'ESTEIRA CAPIRAL'),
(7002, 'ESTEIRA CALCULO'),
(7003, 'ESTEIRA PROVISONAMENTO');
"""

pop_t21 = """
INSERT INTO tbng1021_situ_exeo_flux_capi VALUES
(0, 'ABERTO'),
(1, 'EM ANDAMENTO'),
(2, 'CONCLUIDO');
"""

pop_t4 = """
INSERT INTO tbng1004_agem_flux_capi VALUES
(0, 'TESTE'),
(1, 'VAVA'),
(2, 'SID'),
(3, 'MEIERS'),
(4, 'CIVILIZATION'),
(5, 'AMONG'),
(6, 'US'),
(7, 'GLOBAL'),
(8, 'OFFENSIVE'),
(9, 'VAULT BOY');
"""

pop_t3 = """
INSERT INTO tbng1003_agem_cenr_flux_capi VALUES
(0, 100, 0, 300),
(0, 100, 1, 301),
(0, 100, 2, 302),
(1, 101, 10, 310),
(1, 101, 11, 311),
(1, 101, 12, 312),
(2, 102, 13, 313),
(3, 103, 19, 319),
(3, 103, 20, 320),
(9, 109, 30, 330),
(9, 109, 31, 331);
"""

pop_t13 = """
INSERT INTO tbng1013_cenr_prso_flux_capi VALUES
(0, 300, 0, 500, 2, NULL, NULL),
(0, 300, 1, 511, 0, NULL, NULL),
(1, 301, 2, 522, 0, NULL, NULL),
(1, 301, 3, 533, 1, NULL, NULL),
(10, 310, 4, 544, 2, NULL, NULL),
(11, 311, 5, 555, 2, NULL, NULL),
(13, 313, 8, 588, 1, NULL, NULL),
(19, 319, 6, 566, 0, NULL, NULL),
(20, 320, 7, 577, 0, NULL, NULL),
(30, 330, 8, 593, 2, NULL, NULL),
(30, 330, 9, 594, 2, NULL, NULL),
(31, 331, 10, 595, 2, NULL, NULL);
"""

pop_t99 = """
INSERT INTO tbng1099_arqu_vrto_cenr_prso VALUES
(0, 300, 1, 511, 'v2ser', '2021-04-20',001),
(0, 300, 1, 511, 'v2crd', '2021-04-21',001),
(0, 300, 0, 500, 'rasc', '2021-04-22',002),
(0, 300, 0, 500, 'ggpi', '2021-04-23',002),
(0, 300, 0, 500, 'fato', '2021-04-24',003),
(0, 300, 0, 500, 'rnatc', '2021-04-25',001),
(19, 319, 6, 566, 'bt10c', '2021-04-26',001),
(19, 319, 6, 566, 'bt10d', '2021-04-26',001),
(19, 319, 6, 566, 'bt10e', '2021-04-26',001),
(0, 300, 0, 500, 'v2ser', '2021-04-29',002),
(30, 330, 8, 593, 'kafk1', '2021-04-30',001),
(30, 330, 8, 593, 'kafk3', '2021-04-30',001),
(30, 330, 9, 594, 'kafk4', '2021-04-30',001),
(31, 331, 10, 595, 'kafk7', '2021-04-30',001),
(31, 331, 10, 595, 'kafk9', '2021-04-30',001);
"""

pop_t17 = """
INSERT INTO tbng1017_prso_base_flux_capi VALUES
(0, 500, 90012),
(0, 500, 90013),
(0, 500, 90014),
(1, 511, 90021),
(1, 511, 90022),
(1, 511, 90023),
(5, 555, 90056),
(5, 555, 90057),
(8, 588, 90089);
"""

pop_t2 = """
INSERT INTO tbng1002_vrto_recp_arqu_capi VALUES
('v2ser', 001, '2021-04-20', '2020-12-20', 90012),
('v2ser', 002, '2021-04-22', '2020-12-23', 90013),
('v2ser', 002, '2021-04-29', '2020-12-28', 90014),
('v2crd', 001, '2021-04-21', '2020-12-30', 90015),
('v2crd', 002, '2021-04-29', '2020-12-28', 90016),
('rasc', 002, '2021-04-22', '2020-11-28', 90017),
('bt10c', 001, '2021-01-21', '2020-8-30', 90018),
('bt10d', 002, '2021-01-21', '2020-8-30', 90018),
('bt10e', 001, '2021-01-21', '2020-8-30', 90018),
('bt10c', 001, '2021-04-26', '2020-10-30', 90018),
('bt10d', 002, '2021-04-26', '2020-10-30', 90018),
('bt10e', 001, '2021-04-26', '2020-10-30', 90018),
('kafk1', 001, '2021-04-30', '2020-10-30', 90021),
('kafk3', 001, '2021-04-30', '2020-10-30', 90021),
('kafk4', 001, '2021-04-30', '2020-10-30', 90021),
('kafk7', 001, '2021-04-30', '2020-10-30', 90021),
('kafk9', 001, '2021-04-30', '2020-10-30', 90021);
"""

connection = create_db_connection("localhost", "root", pw, db)
execute_query(connection, pop_t25)
execute_query(connection, pop_t24)
execute_query(connection, pop_t21)
execute_query(connection, pop_t4)
execute_query(connection, pop_t3)
execute_query(connection, pop_t13)
execute_query(connection, pop_t17)
execute_query(connection, pop_t99)
execute_query(connection, pop_t2)

MySQL Database connection successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful
Query successful


## Queries teste

# Programa Real

In [53]:
sql_agendamento = f"""
                    SELECT cod_agem_flux_capi,  num_vers_agem_flux_capi, dat_rfrc_psst
                    FROM tbng1025_vers_agem_flux_capi 
                    WHERE cod_situ_exeo_flux_capi != 2 AND dat_rfrc_psst <= '{date.today().strftime("%Y-%m-%d")}';
                """
lista_agendamentos = list(read_query(connection,sql_agendamento))

for agendamento in lista_agendamentos:
    codigo_agendamento = agendamento[0]
    versao_agendamento = agendamento[1]
    data_agendamento = agendamento[2]
    
    qtd_processos_necessarios = 0
    qtd_processos_concluidos = 0
    print(f'Passando o agendamento de codigo {codigo_agendamento} e a versao {versao_agendamento}')

    sql_cenarios = f"""
                    SELECT cod_cenr_flux_capi, num_vers_cenr_flux_capi
                    FROM tbng1003_agem_cenr_flux_capi
                    WHERE cod_agem_flux_capi = {codigo_agendamento} AND num_vers_agem_flux_capi = {versao_agendamento};
                 """
    lista_cenarios = list(read_query(connection,sql_cenarios))
    qtd_cenarios = list(lista_cenarios)

    for cenario in lista_cenarios:
        print('\n')
        codigo_cenario = cenario[0]
        versao_cenario = cenario[1]
        print(f'\tPassando o cenario de codigo {codigo_cenario} e a versao {versao_cenario}')
        
        sql_processos = f"""
                        SELECT cod_prso_flux_capi, num_vers_prso_flux_capi, cod_situ_exeo_flux_capi
                        FROM tbng1013_cenr_prso_flux_capi
                        WHERE cod_cenr_flux_capi = {codigo_cenario} AND num_vers_cenr_flux_capi = {versao_cenario};
                        """
        lista_processos = list(read_query(connection,sql_processos))

        
        for processo in lista_processos:
            codigo_processo = processo[0]
            versao_processo = processo[1]
            situacao_processo = processo[2]
            print(f'\t\tPassando o processo de codigo {codigo_processo} e a versao {versao_processo}')
            
            qtd_processos_necessarios += 1
            
            #se o processo estiver finalizado não é necessário conferir os arquivos
            if (situacao_processo == 2):
                qtd_processos_concluidos += 1
                continue
            
            if(situacao_processo == 1):
                continue
                
                
            sql_arquivos = f"""
                                SELECT nom_arqu_vrto, dat_rfrc_vrto, num_vrto_arqu
                                FROM tbng1099_arqu_vrto_cenr_prso
                                WHERE cod_cenr_flux_capi = {codigo_cenario} AND num_vers_cenr_flux_capi = {versao_cenario}
                                AND cod_prso_flux_capi = {codigo_processo} AND num_vers_prso_flux_capi = {versao_processo};
                                """
            lista_arquivos = list(read_query(connection,sql_arquivos))
            qtd_arquivos_necessarios = len(lista_arquivos)
            qtd_arquivos_presente = 0
                
            print('\t\t\t\tarquivos necessarios:')
            for l in lista_arquivos:
                print(f'\t\t\t\t{l[0]} {l[1]} {l[2]}')

            if (len(lista_arquivos) <= 0):
                continue
                
            for arquivo in lista_arquivos:
                nome_arquivo = arquivo[0]
                data_arquivo = arquivo[1]
                versao_arquivo = arquivo[2]
                print(f'\t\t\t\tConferindo se o arquivo de nome {nome_arquivo}, versao {versao_arquivo} e data {data_arquivo} existe')

                sql_bate = f"""
                                SELECT *
                                FROM tbng1002_vrto_recp_arqu_capi
                                WHERE nom_arqu_vrto_recp_arqu_capi = "{nome_arquivo}" AND num_vrto_arqu = {versao_arquivo}
                                AND dat_rfrc_vrto_arqu = "{data_arquivo}";
                                """
                
                comparacao = list(read_query(connection,sql_bate))
                        
                if (len(comparacao)>0):
                    print('\t\t\t\t\tO Arquivo existe :)')
                    qtd_arquivos_presente += 1
                else:
                    print('\t\t\t\t\tNÃO EXISTE :(')
                        
            if((qtd_arquivos_necessarios == qtd_arquivos_presente) and qtd_arquivos_necessarios > 0):
                print('\t\t\t\t\t>>>Enviar processo para Glue e atualizar processo para 1')
                #sql_update_processo = f"""
                #                       UPDATE tbng1099_arqu_vrto_cenr_prso
                #                       SET cod_situ_exeo_flux_capi = 1
                #                       WHERE cod_cenr_flux_capi = {codigo_cenario} AND num_vers_cenr_flux_capi = {versao_cenario}
                #                       AND cod_prso_flux_capi = {codigo_processo} AND num_vers_prso_flux_capi = {versao_processo};
                #                       """
                #execute_query(connection, sql_update_processo)
            else:
                print('\t\t\t\t\t>>>Existem arquivos faltando')
    

    if(qtd_processos_concluidos == qtd_processos_necessarios):
        print('agendamento finalizado, parabeins')
        #update = f"""UPDATE tbng1025_vers_agem_flux_capi 
        #           SET cod_situ_exeo_flux_capi = 2
        #           WHERE cod_agem_flux_capi = {codigo_agendamento};"""
        #execute_query(connection, update)
    else:
        print('NAO - agendamento ainda tem processos em aberto')
        
                        
    print('\n-----------------------------')
    
    

Passando o agendamento de codigo 0 e a versao 100


	Passando o cenario de codigo 0 e a versao 300
		Passando o processo de codigo 0 e a versao 500
		Passando o processo de codigo 1 e a versao 511
				arquivos necessarios:
				v2ser 2021-04-20 1
				v2crd 2021-04-21 1
				Conferindo se o arquivo de nome v2ser, versao 1 e data 2021-04-20 existe
					O Arquivo existe :)
				Conferindo se o arquivo de nome v2crd, versao 1 e data 2021-04-21 existe
					O Arquivo existe :)
					>>>Enviar processo para Glue e atualizar processo para 1


	Passando o cenario de codigo 1 e a versao 301
		Passando o processo de codigo 2 e a versao 522
				arquivos necessarios:
		Passando o processo de codigo 3 e a versao 533


	Passando o cenario de codigo 2 e a versao 302
NAO - agendamento ainda tem processos em aberto

-----------------------------
Passando o agendamento de codigo 2 e a versao 102


	Passando o cenario de codigo 13 e a versao 313
		Passando o processo de codigo 8 e a versao 588
NAO - agendamen

In [45]:
connection.close()

In [ ]:
    #update = f"""UPDATE tbng1025_vers_agem_flux_capi 
    #           SET cod_situ_exeo_flux_capi = 1
    #           WHERE cod_agem_flux_capi = {codigo_agendamento};"""
    #execute_query(connection, update)

In [47]:
lista = [[[1,2],[3,4]],[[5,6],[7,8]]]
lista

[[[1, 2], [3, 4]], [[5, 6], [7, 8]]]

In [51]:
lista = [[[1,2],[3,4]],[[5,6],[7,8]]]

for sublista in lista:
    qtd_itens_existentes = 0
    for subsublista in sublista:
        for item in subsublista:
            print(item)

1
2
3
4
5
6
7
8


##### 3.3 - Define Foreign Key Relationships

Now altering the tables to create Foreign Key relationships (see the accompanying [SQL Tutorial Series](https://towardsdatascience.com/tagged/sql-series) on [Towards Data Science](https://towardsdatascience.com/) for the background on all of this), and creating our final table, takes_course

In [8]:
alter_participant = """
ALTER TABLE participant
ADD FOREIGN KEY(client)
REFERENCES client(client_id)
ON DELETE SET NULL;
"""

alter_course = """
ALTER TABLE course
ADD FOREIGN KEY(teacher)
REFERENCES teacher(teacher_id)
ON DELETE SET NULL;
"""

alter_course_again = """
ALTER TABLE course
ADD FOREIGN KEY(client)
REFERENCES client(client_id)
ON DELETE SET NULL;
"""

create_takescourse_table = """
CREATE TABLE takes_course (
  participant_id INT,
  course_id INT,
  PRIMARY KEY(participant_id, course_id),
  FOREIGN KEY(participant_id) REFERENCES participant(participant_id) ON DELETE CASCADE, -- it makes no sense to keep this rtelation when a participant or course is no longer in the system, hence why CASCADE this time
  FOREIGN KEY(course_id) REFERENCES course(course_id) ON DELETE CASCADE
);
"""

connection = create_db_connection("localhost", "root", pw, db)
execute_query(connection, alter_participant)
execute_query(connection, alter_course)
execute_query(connection, alter_course_again)
execute_query(connection, create_takescourse_table)


MySQL Database connection successful
Query successful
Query successful
Query successful
Query successful


-----------------

### 4. Populate Tables

The next step is to populate our tables with data, starting with the teacher table

##### 4.1 - Populate Teacher Table

Here we again assign a multi-line string with our SQL command to a variable, and then call our create_db_connection and execute_query functions.

In [9]:
pop_teacher = """
INSERT INTO teacher VALUES
(1,  'James', 'Smith', 'ENG', NULL, '1985-04-20', 12345, '+491774553676'),
(2, 'Stefanie',  'Martin',  'FRA', NULL,  '1970-02-17', 23456, '+491234567890'), 
(3, 'Steve', 'Wang',  'MAN', 'ENG', '1990-11-12', 34567, '+447840921333'),
(4, 'Friederike',  'Müller-Rossi', 'DEU', 'ITA', '1987-07-07',  45678, '+492345678901'),
(5, 'Isobel', 'Ivanova', 'RUS', 'ENG', '1963-05-30',  56789, '+491772635467'),
(6, 'Niamh', 'Murphy', 'ENG', 'IRI', '1995-09-08',  67890, '+491231231232');
"""

connection = create_db_connection("localhost", "root", pw, db)
execute_query(connection, pop_teacher)

MySQL Database connection successful
Query successful


--------------

### 5. Reading Data

##### 5.1 - Define Data Reading Function

Now that we have populated our tables, it's time to start creating read queries. To do this, we will need a new function.

In [11]:
def read_query(connection, query):
    cursor = connection.cursor()
    result = None
    try:
        cursor.execute(query)
        result = cursor.fetchall()
        return result
    except Error as err:
        print(f"Error: '{err}'")

##### 5.2 - Read Data from Database

Let's try this with a simple query to begin with.

In [12]:
q1 = """
SELECT *
FROM teacher;
"""

connection = create_db_connection("localhost", "root", pw, db)
results = read_query(connection, q1)

for result in results:
  print(result)

MySQL Database connection successful
(1, 'James', 'Smith', 'ENG', None, datetime.date(1985, 4, 20), 12345, '+491774553676')
(2, 'Stefanie', 'Martin', 'FRA', None, datetime.date(1970, 2, 17), 23456, '+491234567890')
(3, 'Steve', 'Wang', 'MAN', 'ENG', datetime.date(1990, 11, 12), 34567, '+447840921333')
(4, 'Friederike', 'Müller-Rossi', 'DEU', 'ITA', datetime.date(1987, 7, 7), 45678, '+492345678901')
(5, 'Isobel', 'Ivanova', 'RUS', 'ENG', datetime.date(1963, 5, 30), 56789, '+491772635467')
(6, 'Niamh', 'Murphy', 'ENG', 'IRI', datetime.date(1995, 9, 8), 67890, '+491231231232')


And here are some more queries to try.

In [13]:
q2 = """
SELECT last_name, dob
FROM teacher;
"""

q3 = """
SELECT *
FROM course
WHERE language = 'ENG'
ORDER BY start_date DESC;
"""

q4 = """
SELECT first_name, last_name, phone_no
FROM teacher
WHERE dob < '1990-01-01';
"""

q5 = """
SELECT course.course_id, course.course_name, course.language, client.client_name, client.address
FROM course
JOIN client
ON course.client = client.client_id
WHERE course.in_school = FALSE;
"""

connection = create_db_connection("localhost", "root", pw, db)
results = read_query(connection, q5)

for result in results:
  print(result)

MySQL Database connection successful
(13, 'Beginner English', 'ENG', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin')
(14, 'Intermediate English', 'ENG', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin')
(15, 'Advanced English', 'ENG', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin')
(17, 'Français intermédiaire', 'FRA', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin')
(19, 'Intermediate English', 'ENG', 'Banko Bank', '12 Betrugstraße, 12345 Berlin')
(20, 'Fortgeschrittenes Russisch', 'RUS', 'AutoMaker AG', '20 Künstlichstraße, 10023 Berlin')


##### 5.3 - Formatting Output into a List

Now we can assign the results to a list, to use further in our python applications or scripts.

The following code returns the results of our query as a list of tuples.

In [14]:
#Initialise empty list
from_db = []

# Loop over the results and append them into our list, different styles

# Returns a list of tuples
for result in results:
  result = result
  from_db.append(result)
    
print(from_db)

[(13, 'Beginner English', 'ENG', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin'), (14, 'Intermediate English', 'ENG', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin'), (15, 'Advanced English', 'ENG', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin'), (17, 'Français intermédiaire', 'FRA', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin'), (19, 'Intermediate English', 'ENG', 'Banko Bank', '12 Betrugstraße, 12345 Berlin'), (20, 'Fortgeschrittenes Russisch', 'RUS', 'AutoMaker AG', '20 Künstlichstraße, 10023 Berlin')]


##### 5.4 - Formatting Output into a List of Lists

If we want to, we can make this return a list of lists instead, like so:

In [15]:
# Returns a list of lists
from_db = []

for result in results:
  result = list(result)
  from_db.append(result)
    
print(from_db)

[[13, 'Beginner English', 'ENG', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin'], [14, 'Intermediate English', 'ENG', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin'], [15, 'Advanced English', 'ENG', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin'], [17, 'Français intermédiaire', 'FRA', 'Big Business Federation', '123 Falschungstraße, 10999 Berlin'], [19, 'Intermediate English', 'ENG', 'Banko Bank', '12 Betrugstraße, 12345 Berlin'], [20, 'Fortgeschrittenes Russisch', 'RUS', 'AutoMaker AG', '20 Künstlichstraße, 10023 Berlin']]


##### 5.5 - Formatting Output into a pandas [DataFrame](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html)

With a little more work (creating a list with our column names), we can create a pandas DataFrame like so:

In [16]:
# Returns a list of lists and then creates a pandas DataFrame
from_db = []

for result in results:
  result = list(result)
  from_db.append(result)


columns = ["course_id", "course_name", "language", "client_name", "address"]
df = pd.DataFrame(from_db, columns=columns)

display(df)

course_id                 course_name language              client_name  \
0         13            Beginner English      ENG  Big Business Federation   
1         14        Intermediate English      ENG  Big Business Federation   
2         15            Advanced English      ENG  Big Business Federation   
3         17      Français intermédiaire      FRA  Big Business Federation   
4         19        Intermediate English      ENG               Banko Bank   
5         20  Fortgeschrittenes Russisch      RUS             AutoMaker AG   

                             address  
0  123 Falschungstraße, 10999 Berlin  
1  123 Falschungstraße, 10999 Berlin  
2  123 Falschungstraße, 10999 Berlin  
3  123 Falschungstraße, 10999 Berlin  
4      12 Betrugstraße, 12345 Berlin  
5   20 Künstlichstraße, 10023 Berlin

### 6. Updating Records

Sometimes we will need to update our Database. We can do this very easily using our execute_query function alongside the SQL [UPDATE](https://dev.mysql.com/doc/refman/8.0/en/update.html) statement.

##### 6.1 - Updating Client Address

The School receives notification that the Big Business Federation has moved office, and now they are located at 23 Fingiertweg, 14534 Berlin. We can change that in our database like so:

In [17]:
update = """
UPDATE client 
SET address = '23 Fingiertweg, 14534 Berlin' 
WHERE client_id = 101;
"""

connection = create_db_connection("localhost", "root", pw, db)
execute_query(connection, update)

MySQL Database connection successful
Query successful


Let's see if that worked.

In [18]:
q1 = """
SELECT *
FROM client
WHERE client_id = 101;
"""

connection = create_db_connection("localhost", "root", pw, db)
results = read_query(connection, q1)

for result in results:
  print(result)

MySQL Database connection successful
(101, 'Big Business Federation', '23 Fingiertweg, 14534 Berlin', 'NGO')


Excellent!

Notice that as well as using "WHERE client_id = 101" in the UPDATE query, we could also have used "WHERE client_name = 'Big Business Federation'" or "WHERE address = '123 Falschungstraße, 10999 Berlin'" (or even "WHERE address LIKE '%Falschung%'"). The important thing is that the WHERE clause allows us to uniquely identify the record we want to update. 

Running the query without a WHERE clause would update all address records in the table, which is very much not what we wanted to do.

-----------------

### 7. Deleting Records

##### 7.1 - Deleting a Course

We can also use our execute_query function to delete records, by using DELETE FROM.

Let's try this with our course table. First let's remind ourselves of the courses contained in the table.

In [19]:
q1 = """
SELECT *
FROM course;
"""

connection = create_db_connection("localhost", "root", pw, db)
results = read_query(connection, q1)

from_db = []

for result in results:
  print(result)

MySQL Database connection successful
(12, 'English for Logistics', 'ENG', 'A1', 10, datetime.date(2020, 2, 1), 1, 1, 105)
(13, 'Beginner English', 'ENG', 'A2', 40, datetime.date(2019, 11, 12), 0, 6, 101)
(14, 'Intermediate English', 'ENG', 'B2', 40, datetime.date(2019, 11, 12), 0, 6, 101)
(15, 'Advanced English', 'ENG', 'C1', 40, datetime.date(2019, 11, 12), 0, 6, 101)
(16, 'Mandarin für Autoindustrie', 'MAN', 'B1', 15, datetime.date(2020, 1, 15), 1, 3, 103)
(17, 'Français intermédiaire', 'FRA', 'B1', 18, datetime.date(2020, 4, 3), 0, 2, 101)
(18, 'Deutsch für Anfänger', 'DEU', 'A2', 8, datetime.date(2020, 2, 14), 1, 4, 102)
(19, 'Intermediate English', 'ENG', 'B2', 10, datetime.date(2020, 3, 29), 0, 1, 104)
(20, 'Fortgeschrittenes Russisch', 'RUS', 'C1', 4, datetime.date(2020, 4, 8), 0, 5, 103)


Let's delete the course with course_id 20 - 'Fortgeschrittenes Russisch'. For this we will use the [DELETE](https://dev.mysql.com/doc/refman/8.0/en/delete.html) SQL command.

In [20]:
delete_course = """
DELETE FROM course WHERE course_id = 20;
"""

connection = create_db_connection("localhost", "root", pw, db)
execute_query(connection, delete_course)

MySQL Database connection successful
Query successful


Let's confirm that the course is gone.

In [21]:
q1 = """
SELECT *
FROM course;
"""

connection = create_db_connection("localhost", "root", pw, db)
results = read_query(connection, q1)

from_db = []

for result in results:
  print(result)

MySQL Database connection successful
(12, 'English for Logistics', 'ENG', 'A1', 10, datetime.date(2020, 2, 1), 1, 1, 105)
(13, 'Beginner English', 'ENG', 'A2', 40, datetime.date(2019, 11, 12), 0, 6, 101)
(14, 'Intermediate English', 'ENG', 'B2', 40, datetime.date(2019, 11, 12), 0, 6, 101)
(15, 'Advanced English', 'ENG', 'C1', 40, datetime.date(2019, 11, 12), 0, 6, 101)
(16, 'Mandarin für Autoindustrie', 'MAN', 'B1', 15, datetime.date(2020, 1, 15), 1, 3, 103)
(17, 'Français intermédiaire', 'FRA', 'B1', 18, datetime.date(2020, 4, 3), 0, 2, 101)
(18, 'Deutsch für Anfänger', 'DEU', 'A2', 8, datetime.date(2020, 2, 14), 1, 4, 102)
(19, 'Intermediate English', 'ENG', 'B2', 10, datetime.date(2020, 3, 29), 0, 1, 104)


It's gone! Good work.

##### 7.2 - Restoring the Course

Let's put that course back - it's a perfectly good course.

In [22]:
restore_russian = """
INSERT INTO course VALUES
(20, 'Fortgeschrittenes Russisch', 'RUS', 'C1',  4, '2020-04-08',  FALSE, 5, 103);
"""

connection = create_db_connection("localhost", "root", pw, db)
execute_query(connection, restore_russian)


q1 = """
SELECT *
FROM course;
"""

connection = create_db_connection("localhost", "root", pw, db)
results = read_query(connection, q1)

from_db = []

for result in results:
  print(result)

MySQL Database connection successful
Query successful
MySQL Database connection successful
(12, 'English for Logistics', 'ENG', 'A1', 10, datetime.date(2020, 2, 1), 1, 1, 105)
(13, 'Beginner English', 'ENG', 'A2', 40, datetime.date(2019, 11, 12), 0, 6, 101)
(14, 'Intermediate English', 'ENG', 'B2', 40, datetime.date(2019, 11, 12), 0, 6, 101)
(15, 'Advanced English', 'ENG', 'C1', 40, datetime.date(2019, 11, 12), 0, 6, 101)
(16, 'Mandarin für Autoindustrie', 'MAN', 'B1', 15, datetime.date(2020, 1, 15), 1, 3, 103)
(17, 'Français intermédiaire', 'FRA', 'B1', 18, datetime.date(2020, 4, 3), 0, 2, 101)
(18, 'Deutsch für Anfänger', 'DEU', 'A2', 8, datetime.date(2020, 2, 14), 1, 4, 102)
(19, 'Intermediate English', 'ENG', 'B2', 10, datetime.date(2020, 3, 29), 0, 1, 104)
(20, 'Fortgeschrittenes Russisch', 'RUS', 'C1', 4, datetime.date(2020, 4, 8), 0, 5, 103)


Excellent.

This also works with DROP TABLE if we want to get rid of a whole table at once. We won't do that in this tutorial, but feel free to try it yourself. In this notebook we can always create and populate the table again - in a real production environment we will need to be much more careful with the DELETE and DROP operations. 

And always remember to back up our database!

----------------

### 8. Creating Records from Lists

We saw in Section 4 that we can use the SQL [INSERT](https://dev.mysql.com/doc/refman/8.0/en/insert.html) command in our execute_query function to insert records into our tables.

MySQL Connector also gives us a way to do this in a more 'pythonic' fashion, using a list of tuples as our input, where each tuple contains the data we wish to insert into our table. This is extremely useful for updating our database with data which may have been generated by an application we have written in Python, such as logs of user activity on a social media app, for example.

To do this, we will use the [executemany()](https://dev.mysql.com/doc/connector-python/en/connector-python-api-mysqlcursor-executemany.html) method, instead of the simpler [execute()](https://dev.mysql.com/doc/connector-python/en/connector-python-api-mysqlcursor-execute.html) method we have been using thus far.

This method is also more secure if our database is open to our users at any point, as it helps to prevent against [SQL Injection](https://en.wikipedia.org/wiki/SQL_injection) attacks, which can damage or even destroy our whole database.

##### 8.1 - Create Execute List Query Function

To see how we can do this, let's add a couple of new teachers to our teacher table.

Fist let's modify our execute_query function to use [executemany()](https://dev.mysql.com/doc/connector-python/en/connector-python-api-mysqlcursor-executemany.html) and to accept one more argument.

In [23]:
def execute_list_query(connection, sql, val):
    cursor = connection.cursor()
    try:
        cursor.executemany(sql, val)
        connection.commit()
        print("Query successful")
    except Error as err:
        print(f"Error: '{err}'")

##### 8.2 - Add New Teachers

Now let's create a list containing the data for our new teachers (each stored within a [tuple](https://www.w3schools.com/python/python_tuples.asp)), and the SQL command to perform our action.

Notice that the SQL command requires a '%s' placeholder for each of the columns we wish to act upon, so in this case we need 8 for the 8 columns we wish to add values for.

In [24]:
sql = '''
    INSERT INTO teacher (teacher_id, first_name, last_name, language_1, language_2, dob, tax_id, phone_no) 
    VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
    '''
val = [
    (7, 'Hank', 'Dodson', 'ENG', None, '1991-12-23', 11111, '+491772345678'), 
    (8, 'Sue', 'Perkins', 'MAN', 'ENG', '1976-02-02', 22222, '+491443456432')
]


connection = create_db_connection("localhost", "root", pw, db)
execute_list_query(connection, sql, val)

MySQL Database connection successful
Query successful


In [25]:
q1 = """
SELECT *
FROM teacher;
"""

connection = create_db_connection("localhost", "root", pw, db)
results = read_query(connection, q1)

from_db = []

for result in results:
  print(result)

MySQL Database connection successful
(1, 'James', 'Smith', 'ENG', None, datetime.date(1985, 4, 20), 12345, '+491774553676')
(2, 'Stefanie', 'Martin', 'FRA', None, datetime.date(1970, 2, 17), 23456, '+491234567890')
(3, 'Steve', 'Wang', 'MAN', 'ENG', datetime.date(1990, 11, 12), 34567, '+447840921333')
(4, 'Friederike', 'Müller-Rossi', 'DEU', 'ITA', datetime.date(1987, 7, 7), 45678, '+492345678901')
(5, 'Isobel', 'Ivanova', 'RUS', 'ENG', datetime.date(1963, 5, 30), 56789, '+491772635467')
(6, 'Niamh', 'Murphy', 'ENG', 'IRI', datetime.date(1995, 9, 8), 67890, '+491231231232')
(7, 'Hank', 'Dodson', 'ENG', None, datetime.date(1991, 12, 23), 11111, '+491772345678')
(8, 'Sue', 'Perkins', 'MAN', 'ENG', datetime.date(1976, 2, 2), 22222, '+491443456432')
